index.d.ts

```diff
@@ -746,11 +746,23 @@ declare namespace Moleculer {
 		statuses: Array<{ name: string; available: boolean }>;
 	}
 
-	class Service<S = ServiceSettingSchema> implements ServiceSchema {
-		constructor(broker: ServiceBroker, schema?: ServiceSchema<S>);
-
+	class Service<S = ServiceSettingSchema, SM = ServiceSettingSchema> implements ServiceSchema {
```

Comment:
Could you explain why `SM` needs?

Reply:
In case you want to have a certain definition of the schema mod settings, you can strictly type the schema mods. Is it very useful? Not really, but i do believe that people should have at least the ability to enforce the schema mod structure.

Reply:
@shawnmcknight what do you think?

Reply:
The main reason for the `S = ServiceSettingSchema` generic was to allow `this.settings` to be typed so it wasn't basically `Record<string, any>`.  It wasn't really to gain any sort of type safety in the `Service` constructor, although it would appear that ended up as a side effect since the `schema` parameter is typed.  I would suggest that if you're trying to merge schemas with disparate `settings` properties via providing `schemaMods` then this is going to turn out badly because you either won't get the proper type of the `schema` parameter or the wrong type for the `settings` property since they use the same generic.

Is there any external use case for `schemaMods` being provided to the constructor?  That is, for what reason would an external consumer provide this parameter to the `Service` constructor.  I do see it documented in the `ServiceBroker` API and it looks like it was recently added when the static members of `Service` were made instance members.  What is the use case for this?  Is there one for which it makes sense to expose this as a parameter for external consumption?

Reply:
@intech could you explain the new `schemaMods` parameter of service constructor?

Reply:
> The main reason for the `S = ServiceSettingSchema` generic was to allow `this.settings` to be typed so it wasn't basically `Record<string, any>`. It wasn't really to gain any sort of type safety in the `Service` constructor, although it would appear that ended up as a side effect since the `schema` parameter is typed. I would suggest that if you're trying to merge schemas with disparate `settings` properties via providing `schemaMods` then this is going to turn out badly because you either won't get the proper type of the `schema` parameter or the wrong type for the `settings` property since they use the same generic.
> 
> Is there any external use case for `schemaMods` being provided to the constructor? That is, for what reason would an external consumer provide this parameter to the `Service` constructor. I do see it documented in the `ServiceBroker` API and it looks like it was recently added when the static members of `Service` were made instance members. What is the use case for this? Is there one for which it makes sense to expose this as a parameter for external consumption?

Seems like the main use case for it seems to be testing. The typescript starter includes some tests where schemaMods is used in order to override an action so it could be checked using jest. As for more practical use cases... not sure. Changing it would be a breaking change though.

Reply:
@icebob It has been ported from [service-broker](https://github.com/moleculerjs/moleculer/pull/1098/files#diff-3cd4f90d2dcbb1220a816dee13337a1f0d6853c2da7b5bf61bf10c68ce62e43aL827-L830) to [service](https://github.com/moleculerjs/moleculer/pull/1098/files#diff-145d31b275fbdb6a7b17ef982f83babd2323896502f2c8374da313a8833e0ebaR51-R57) to preserve the ability to call the `mergeSchemas` method inside the instance.
I seem to have seen its use only in tests.

Reply:
Not to be annoying, but if this issue is not being discussed, and there isn't any other feedback provided, could we look into getting the MR merged?

Not having these types fixed means typescript is broken on this and upcoming versions as long for anyone who is using any of the schema merge methods.

Reply:
So, I've checked the source and figured out the root of the `schemaMods`. It's a very old "feature", it was implemented in early versions [6 years ago](https://github.com/moleculerjs/moleculer/commit/6628763e93bcf86fb650bca3b31caf2820db619e) when the `mixins` logic is not implemented yet. So this parameter was added to extend an existing service schema. 
But since the mixins feature, it's not used, but we cannot remove it because it can cause breaking change if somebody still uses it. So I removed it from the d.ts, but kept it in the code and added a deprecation warning. So we will able to remove it in version 0.15.

Moreover, I've removed from the PR the extra exposed internal methods. I don't like if we expose the internal methods because the users can use them directly because it will cause problems when we refactor them.